### DISCUSSION TITLE
SDDEBUG to OpenLOG

### DISCUSSION THREAD

**Discussion by RobertDeHate** [qqqlab/madflight#33] [Q&A]
I found in madflight_board_default_RP2040.h
//Serial debug on USB Serial port (no GPIO pins)
Is there a way to output the BB data to a serial port so I can off load data writing to an Open Log board.
When I implement bb.log_imu(); it will lock up the controller.
I assume it's due to the write time required for the SD so I would like to write it to the Open Log serial.

Thanks,
Robert DeHate

**Comment by qqqlab** [comment]
Logging to openlog is something I wanted to do, but did not get to it yet.

However, I tested 1000Hz full speed IMU logging on a RP2040 (Raspberry Pi Pico) with a sdcard breakout and it worked fine.

Does logging work when you don't use bb.log_imu()?

Please post your code changes and give a short board/hardware description, maybe I can help to get a step further.

**Reply by RobertDeHate** [reply]
Logging works fine w/o bb.log_imu()
It usually runs about 2 minutes before it crashes.
The CLI stops responding and the LED stops, sometimes on, sometimes off.
Tried multiple card types from 128M, 1G, 8G, 32G.
I am using a RPi 2 w/RP2350, IMU MPU9350, SD reader off amazon https://tinyurl.com/djcuf2fe
Not sure how it knows I have a 2350 instead of a 2040
I got OpenLog to work and simply swiped the RC UART as I dont have a radio hooked up because I am just capturing data.
I hacked it in modifying hw_RP2040.h bb_sdcard.h and madflight.h
```
Changes for UART BBDebug

FILE hw_RP2040.h

//-------------------------------------
//Bus Setup
//-------------------------------------

SerialUART Debug_Serial = SerialUART(uart0, 1, 0 );

#if defined(HW_PIN_GPS_TX) && defined(HW_PIN_GPS_RX)
  // GPS Serial - uncomment one: SerialUART, SerialIRQ or SerialPIO and use uart0 or uart1
  uint8_t gps_txbuf[256];
  uint8_t gps_rxbuf[256];
  SerialIRQ gps_Serial(uart1, HW_PIN_GPS_TX, gps_txbuf, sizeof(gps_txbuf), HW_PIN_GPS_RX, gps_rxbuf, sizeof(gps_rxbuf)); //SerialIRQ uses buffered tx and rx in addition to the 32 byte uart hardware fifo
  //SerialUART gps_Serial = SerialUART(uart1, HW_PIN_GPS_TX, HW_PIN_GPS_RX); //SerialUART default Arduino impementation (had some problems with this)
  //SerialPIO gps_Serial = SerialPIO(HW_PIN_GPS_TX, HW_PIN_GPS_RX, 32); //PIO uarts, any pin allowed (not tested)
#else
  //DUMMY
  SerialUART gps_Serial = SerialUART(uart1, -1, -1);
#endif

#if defined(HW_PIN_RCIN_TX) && defined(HW_PIN_RCIN_RX)
  // RC Input Serial - uncomment one: SerialUART, SerialIRQ or SerialPIO and use uart0 or uart1
  uint8_t rcin_txbuf[256];
  uint8_t rcin_rxbuf[1024];
  SerialIRQ *rcin_Serial = new SerialIRQ(uart1, HW_PIN_RCIN_TX, rcin_txbuf, sizeof(rcin_txbuf), HW_PIN_RCIN_RX, rcin_rxbuf, sizeof(rcin_rxbuf)); //SerialIRQ uses buffered tx and rx in addition to the 32 byte uart hardware fifo
  //SerialUART rcin_Serial = SerialUART(uart0, HW_PIN_RCIN_TX, HW_PIN_RCIN_RX); //SerialUART default Arduino impementation (had some problems with this)
  //SerialPIO *rcin_Serial = new SerialPIO(HW_PIN_RCIN_TX, HW_PIN_RCIN_RX, 32); //PIO uarts, any pin allowed (not tested)
#else
  //DUMMY
  SerialUART *rcin_Serial = new SerialUART(uart1, -1, -1);
#endif

FILE bb_sdcard.h

//=====================================================================================================================
// Logging to Serial
//=====================================================================================================================
#if BB_USE == BB_USE_SDDEBUG

class BBFS_Debug : public BBFS {
public:
  void setup() override {}
  bool writeOpen() override {return true;}
  
  void write(const uint8_t *buf, const uint8_t len) override {
    Debug_Serial.print("BinLog:");
    for(int i=0;i<len;i++) Debug_Serial.printf("%02X ", buf[i]);
    Debug_Serial.println();
  }

FILE madflight.h

  //===============================================================================================
  // madflight_setup()
  //===============================================================================================

  void madflight_setup() {
    led.setup(); //Set built in LED to turn on to signal startup
    Serial.begin(115200); //start console serial
    Debug_Serial.begin(115200); //start Debug serial

```

Yes I probably broke GPS and RC making them the same UART. But not using at this time.
But still have to figure out the best way to convert ascii encoded HEX to integer values w/o writing my own program.
It is easier to use OpenLog as it only requires 3 wires and isn't much bigger than an SD card reader.
Also I read some concerns Drhem flight had about SD card write times.

Robert

**Reply by qqqlab** [reply]
>It usually runs about 2 minutes before it crashes.

Are you using madflight version 1.3.0 or later? I had this problem in an older version.

>concerns Drhem flight had about SD card write times

That's why I started to use FreeRTOS: loop() and sdcard run as separate tasks with same low priority. If the sdcard writes are too slow, then you will have dropped records, but loop() is not blocked..

**Reply by robert-b** [reply]
utilize dma and there is no need for free rtos
